platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,200 @@
+        RosterValidator.validate(candidateRoster);
+
+        final Bytes incomingCandidateRosterHash =
+                RosterUtils.hash(candidateRoster).getBytes();
+
+        // update the roster state/map
+        final RosterState previousRosterState = rosterStateOrThrow();
+        final Bytes previousCandidateRosterHash = previousRosterState.candidateRosterHash();
+        final Builder newRosterState =
+                previousRosterState.copyBuilder().candidateRosterHash(incomingCandidateRosterHash);
+        removeRoster(previousCandidateRosterHash);
+        storeRoster(candidateRoster, incomingCandidateRosterHash, newRosterState);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getCandidateRoster() {
+        return rosterStateAccessor.getCandidateRoster();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getActiveRoster() {
+        return rosterStateAccessor.getActiveRoster();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public void setActiveRoster(@NonNull final Roster roster, final long round) {
+        Objects.requireNonNull(roster);
+        RosterValidator.validate(roster);
+
+        // update the roster state
+        final RosterState previousRosterState = rosterStateOrThrow();
+        final List<RoundRosterPair> roundRosterPairs = new LinkedList<>(previousRosterState.roundRosterPairs());
+        final Bytes activeRosterHash = RosterUtils.hash(roster).getBytes();
+        if (!roundRosterPairs.isEmpty()) {
+            final RoundRosterPair activeRosterPair = roundRosterPairs.getFirst();
+            if (round < 0 || round <= activeRosterPair.roundNumber()) {
+                throw new IllegalArgumentException(
+                        "incoming round number must be greater than the round number of the current active roster.");
+            }
+        }
+        roundRosterPairs.addFirst(new RoundRosterPair(round, activeRosterHash));
+
+        if (roundRosterPairs.size() > MAXIMUM_ROSTER_HISTORY_SIZE) {
+            final RoundRosterPair lastRemovedRoster = roundRosterPairs.removeLast();
+            removeRoster(lastRemovedRoster.activeRosterHash());
+
+            // At this phase of the implementation, the roster state has a fixed size limit for active rosters.
+            // Future implementations (e.g. DAB) can modify this.
+            if (roundRosterPairs.size() > MAXIMUM_ROSTER_HISTORY_SIZE) {
+                // additional safety check to ensure that the roster state does not contain more than set limit.
+                throw new IllegalStateException(
+                        "Active rosters in the Roster state cannot be more than  " + MAXIMUM_ROSTER_HISTORY_SIZE);
+            }
+        }
+
+        final Builder newRosterState = previousRosterState
+                .copyBuilder()
+                .candidateRosterHash(previousRosterState.candidateRosterHash())
```

Comment:
previously, we were setting the candidate roster hash to empty because we were adopting it. Now, when we set an active roster, we simply add it to the roster state/map and leave existing candidate roster as-is, if any. The code to adopt, which is yet to be done, will do this setting to empty once again when done.

Reply:
This means there would be more state commits than necessary: one to set the active roster, and then another one to re-set the roster state with an empty candidate roster.

Reply:
The APIs provided by this class are now fairly independent. They will be called by different components at different times - TSS Base Service calling setCandidateRoster and ServicesMain calling setActiveRoster etc. When the adopt API is introduced, it too will need to make the appropriate state manipulations and commit (which will include the code to set the hash to empty). The independent state commits seem an acceptable trade-off for this flexibility.

Reply:
Won't this cause the roster state to be written to the block stream twice when adopting a candidate roster? Since there are multiple committed modifications?

IMO, the standard case suffering a performance penalty for the sake of flexibility (which I'm not convinced we need in this case) doesn't make sense.

Reply:
Can you clarify how it will be written to the block stream twice? That code to adopt doesn't exist yet. When it's introduced, it will likely share most of the code setActiveRoster uses with the additional operation to unset the `candidateRosterHash`.

Reply:
I made my statement under the assumption that the code to adopt the candidate roster would just leverage the methods created in this PR. If that were to be the case, it would involve a call to `setActiveRoster`, followed by a call to `setCandidateRoster`. Hence "multiple committed modifications".

Are you saying that a new method `adoptCandidateRoster` will be added to `WritableRosterStore`, and the adoption wouldn't use `setActiveRoster` and `setCandidateRoster`?

Reply:
Thinking through the use cases, I think we have:

1. On upgrade, the candidate roster is set as the active roster.
2. On genesis startup, a genesis roster is provided and is set as the active roster.
3. On non-genesis startup as part of a network transplant, an override roster is provided and is set as the active roster.

In all these cases, I think it is desirable to clear out the candidate roster hash. For 1, the candidate roster is being set as active and is no longer a candidate. For 2, there is no candidate roster so no harm in clearing it. For 3, we definitely want to clear the candidate roster because if there is one, it is for a different network.

There was a push to make the operations in this class much more like traditional setter/getters, but given our use cases, I think it makes sense to set the candidate roster has to empty bytes here. It is desirable in each case and it reduces the number of changes to the block stream.

Reply:
remove = get + delete?

Reply:
to be clear, this logic preserves the candidate roster while setting an active roster.
Setting `candidateRosterHash` here to empty byte is effectively unsetting the candidate roster in the state.
Do we really want to unset a candidate roster from the state whilst setting an active roster?

Reply:
> 3. On non-genesis startup as part of a network transplant, an override roster is provided and is set as the active roster.

Is it really necessary to implement this code pathway? A standalone tool could directly modify a state-to-be-transplanted, so that it contains the desired roster. That way, we wouldn't have what essentially amounts to test code as part of production.

Reply:
@litt3, can we have the discussion about the override roster behavior separately?

Reply:
+1 on clearing the candidate roster bytes.

Reply:
discussed offline - the existing candidate roster must be removed when setting active roster. Updated.

Reply:
I see code setting the candidate roster hash to empty bytes, but i dont see code to remove it from the roster map if there are no remaining references to its hash in the roundroster pair list.

Reply:
it's on line 156 of the current version just before it's stored.

Reply:
my mistake - thank you @kfa-aguda